File: src/core/models/TargetModel.js
// 目标运动学模型
import VectorUtils from '../utils/VectorUtils.js';

class TargetModel {
  constructor(params) {
    this.position = params.position || [5000, 3000, 1000];
    this.velocity = params.velocity || [-200, 0, 0];
    this.acceleration = params.acceleration || [0, 0, 0];
    
    this.motionType = params.motionType || 'constant';
    this.motionParams = params.motionParams || {};
    
    this.time = 0;
    this.previousAccel = [0, 0, 0]; // 用于平滑随机机动
    this.zigzagDirection = 1; // 用于之字形机动
    this.zigzagSwitchTime = 0; // 用于之字形机动
  }

  // 更新目标状态（增强机动模式）
  update(dt) {
    this.time += dt;
    
    // 根据运动类型更新加速度
    switch (this.motionType) {
      case 'constant':
        // 匀速直线运动
        this.acceleration = [0, 0, 0];
        break;
        
      case 'circular':
        // 圆周运动（改进版：绕指定中心点）
        const omega = this.motionParams.omega || 0.1;
        const center = this.motionParams.center || [0, 0, 0];
        
        // 计算相对于中心点的位置
        const relPos = this.position.map((val, i) => val - center[i]);
        const radius = VectorUtils.norm(relPos);
        
        // 圆周运动加速度 = -ω² * 相对位置（指向圆心）
        this.acceleration = relPos.map(val => -omega * omega * val);
        break;
        
      case 'sine':
        // 正弦机动（3D支持）
        const amplitude = this.motionParams.amplitude || 100;
        const frequency = this.motionParams.frequency || 0.5;
        const phase = this.motionParams.phase || 0;
        
        // 在y和z方向生成正弦加速度
        this.acceleration[0] = 0;
        this.acceleration[1] = -amplitude * frequency * frequency * Math.sin(frequency * this.time + phase);
        this.acceleration[2] = -amplitude * frequency * frequency * Math.sin(frequency * this.time + phase + Math.PI/2);
        break;
        
      case 'random':
        // 随机运动（平滑版）
        const maxAccel = this.motionParams.maxAccel || 50;
        const smoothFactor = this.motionParams.smoothFactor || 0.3;
        
        // 生成随机加速度
        const randomAccel = [
          (Math.random() - 0.5) * 2 * maxAccel,
          (Math.random() - 0.5) * 2 * maxAccel,
          (Math.random() - 0.5) * 2 * maxAccel
        ];
        
        // 平滑处理
        this.acceleration = this.acceleration.map((val, i) => 
          val * (1 - smoothFactor) + randomAccel[i] * smoothFactor
        );
        
        this.previousAccel = [...this.acceleration];
        break;
        
      case 'evasive':
        // 智能规避运动
        const evasionFreq = this.motionParams.evasionFreq || 1.0;
        const evasionAmplitude = this.motionParams.evasionAmplitude || 100;
        
        // 生成垂直于速度方向的机动
        const speed = VectorUtils.norm(this.velocity);
        const velocityDir = speed > 0 ? VectorUtils.normalize(this.velocity) : [1, 0, 0];
        
        // 生成垂直于速度方向的单位向量
        const perpendicular1 = [0, velocityDir[2], -velocityDir[1]];
        const perpendicular2 = VectorUtils.cross(velocityDir, perpendicular1);
        
        // 在垂直于速度方向上生成正弦机动
        this.acceleration = VectorUtils.add(
          VectorUtils.multiply(perpendicular1, evasionAmplitude * Math.sin(evasionFreq * this.time)),
          VectorUtils.multiply(perpendicular2, evasionAmplitude * Math.cos(evasionFreq * this.time))
        );
        break;
        
      case 'zigzag':
        // 之字形机动
        const zigzagAmplitude = this.motionParams.zigzagAmplitude || 100;
        const zigzagFreq = this.motionParams.zigzagFreq || 1.0;
        const switchInterval = 1 / zigzagFreq;
        
        // 定期切换方向
        if (this.time >= this.zigzagSwitchTime) {
          this.zigzagDirection *= -1;
          this.zigzagSwitchTime = this.time + switchInterval;
        }
        
        // 生成之字形加速度
        this.acceleration = [
          0,
          this.zigzagDirection * zigzagAmplitude,
          0
        ];
        break;
        
      case 'spiral':
        // 螺旋机动
        const spiralFreq = this.motionParams.spiralFreq || 0.5;
        const spiralAmplitude = this.motionParams.spiralAmplitude || 50;
        const spiralGrowth = this.motionParams.spiralGrowth || 0.1;
        
        // 生成螺旋加速度
        const spiralRadius = spiralGrowth * this.time;
        this.acceleration[0] = -spiralFreq * spiralFreq * spiralRadius * Math.cos(spiralFreq * this.time);
        this.acceleration[1] = -spiralFreq * spiralFreq * spiralRadius * Math.sin(spiralFreq * this.time);
        this.acceleration[2] = spiralAmplitude * Math.sin(spiralFreq * this.time);
        break;
        
      default:
        this.acceleration = [0, 0, 0];
    }
    
    // 更新速度
    this.velocity = this.velocity.map((v, i) => v + this.acceleration[i] * dt);
    
    // 更新位置
    this.position = this.position.map((p, i) => p + this.velocity[i] * dt);
    
    return this;
  }

  // 获取当前状态
  getState() {
    return {
      position: [...this.position],
      velocity: [...this.velocity],
      acceleration: [...this.acceleration]
    };
  }

  // 重置状态
  reset(params) {
    this.position = params.position || [5000, 3000, 1000];
    this.velocity = params.velocity || [-200, 0, 0];
    this.acceleration = params.acceleration || [0, 0, 0];
    this.time = 0;
    this.previousAccel = [0, 0, 0];
    this.zigzagDirection = 1;
    this.zigzagSwitchTime = 0;
    return this;
  }
}

export default TargetModel

File: models/TargetModel.m
classdef TargetModel < handle
    properties
        position
        velocity
        acceleration
        motion_type
        motion_params
        time
    end
    
    methods
        function obj = TargetModel(initial_state, motion_type, motion_params)
            if nargin < 3
                motion_params = struct();
            end
            if nargin < 2
                motion_type = 'constant';
            end
            
            obj.position = initial_state.position;
            obj.velocity = initial_state.velocity;
            obj.acceleration = initial_state.acceleration;
            obj.motion_type = motion_type;
            obj.motion_params = motion_params;
            obj.time = 0;
        end
        
        function obj = update(obj, dt)
            obj.time = obj.time + dt;
            
            switch obj.motion_type
                case 'constant'
                    obj = update_constant(obj, dt);
                case 'circular'
                    obj = update_circular(obj, dt);
                case 'sine'
                    obj = update_sine(obj, dt);
                case 'random'
                    obj = update_random(obj, dt);
                case 'evasive'
                    obj = update_evasive(obj, dt);
                otherwise
                    obj = update_constant(obj, dt);
            end
            
            obj.position = obj.position + obj.velocity * dt;
        end
        
        function obj = update_constant(obj, dt)
            obj.acceleration = [0; 0; 0];
        end
        
        function obj = update_circular(obj, dt)
            if isfield(obj.motion_params, 'turn_rate')
                turn_rate = obj.motion_params.turn_rate;
            else
                turn_rate = 0.5;
            end
            
            speed = norm(obj.velocity);
            if speed > 1e-6
                velocity_dir = obj.velocity / speed;
                
                R = [cos(turn_rate*dt), -sin(turn_rate*dt), 0;
                     sin(turn_rate*dt), cos(turn_rate*dt), 0;
                     0, 0, 1];
                
                new_velocity_dir = R * velocity_dir;
                obj.velocity = new_velocity_dir * speed;
                obj.acceleration = (new_velocity_dir - velocity_dir) * speed / dt;
            end
        end
        
        function obj = update_sine(obj, dt)
            if isfield(obj.motion_params, 'amplitude')
                amplitude = obj.motion_params.amplitude;
            else
                amplitude = 50;
            end
            
            if isfield(obj.motion_params, 'frequency')
                frequency = obj.motion_params.frequency;
            else
                frequency = 0.5;
            end
            
            speed = norm(obj.velocity);
            if speed > 1e-6
                velocity_dir = obj.velocity / speed;
                
                lateral_accel = amplitude * sin(2 * pi * frequency * obj.time);
                
                if abs(velocity_dir(1)) > abs(velocity_dir(2))
                    perp_dir = [0; 1; 0];
                else
                    perp_dir = [1; 0; 0];
                end
                
                obj.acceleration = perp_dir * lateral_accel;
            end
        end
        
        function obj = update_random(obj, dt)
            if isfield(obj.motion_params, 'max_accel')
                max_accel = obj.motion_params.max_accel;
            else
                max_accel = 20;
            end
            
            if isfield(obj.motion_params, 'change_prob')
                change_prob = obj.motion_params.change_prob;
            else
                change_prob = 0.1;
            end
            
            if rand() < change_prob
                obj.acceleration = (rand(3, 1) - 0.5) * 2 * max_accel;
            end
        end
        
        function obj = update_evasive(obj, dt)
            if isfield(obj.motion_params, 'turn_rate')
                turn_rate = obj.motion_params.turn_rate;
            else
                turn_rate = 0.8;
            end
            
            if isfield(obj.motion_params, 'switch_time')
                switch_time = obj.motion_params.switch_time;
            else
                switch_time = 2.0;
            end
            
            speed = norm(obj.velocity);
            if speed > 1e-6
                velocity_dir = obj.velocity / speed;
                
                current_turn = turn_rate * sin(2 * pi * obj.time / switch_time);
                
                R = [cos(current_turn*dt), -sin(current_turn*dt), 0;
                     sin(current_turn*dt), cos(current_turn*dt), 0;
                     0, 0, 1];
                
                new_velocity_dir = R * velocity_dir;
                obj.velocity = new_velocity_dir * speed;
                obj.acceleration = (new_velocity_dir - velocity_dir) * speed / dt;
            end
        end
        
        function state = get_state(obj)
            state.position = obj.position;
            state.velocity = obj.velocity;
            state.acceleration = obj.acceleration;
            state.speed = norm(obj.velocity);
        end
        
        function obj = set_state(obj, state)
            if isfield(state, 'position')
                obj.position = state.position;
            end
            if isfield(state, 'velocity')
                obj.velocity = state.velocity;
            end
            if isfield(state, 'acceleration')
                obj.acceleration = state.acceleration;
            end
        end
        
        function reset(obj, initial_state)
            obj.position = initial_state.position;
            obj.velocity = initial_state.velocity;
            obj.acceleration = initial_state.acceleration;
            obj.time = 0;
        end
    end
end
